installinator/src/dispatch.rs

```diff
@@ -184,54 +179,121 @@ impl InstallOpts {
+                InstallinatorStepId::Download,
+                "Downloading artifact",
+                |cx| async move {
+                    let host_phase_2_artifact =
+                        fetch_artifact(&cx, &host_phase_2_id, discovery, &log)
+                            .await?;
+
+                    let address = host_phase_2_artifact.addr;
+
+                    StepResult::success(
+                        host_phase_2_artifact,
+                        InstallinatorCompletionMetadata::Download { address },
+                    )
+                },
+            )
+            .register();
 
         let control_plane_id = ArtifactHashId {
             kind: KnownArtifactKind::ControlPlane.into(),
             hash: image_id.control_plane,
         };
-
-        let control_plane_artifact = fetch_artifact(
-            &control_plane_id,
-            &self.discover_opts.mechanism,
-            &log,
-            &event_sender,
-        )
-        .await?;
+        let control_plane_id_2 = control_plane_id.clone();
+        let control_plane_artifact = engine
+            .new_step(
+                InstallinatorComponent::ControlPlane,
+                InstallinatorStepId::Download,
+                "Downloading artifact",
+                |cx| async move {
+                    let control_plane_artifact =
+                        fetch_artifact(&cx, &control_plane_id, discovery, &log)
+                            .await?;
+
+                    let address = control_plane_artifact.addr;
+
+                    StepResult::success(
+                        control_plane_artifact,
+                        InstallinatorCompletionMetadata::Download { address },
+                    )
+                },
+            )
+            .register();
 
         let destination = if self.install_on_gimlet {
             WriteDestination::from_hardware(&log)?
         } else {
-            // clap ensures `self.destinatino` is not `None` if
+            // clap ensures `self.destination` is not `None` if
             // `install_on_gimlet` is false.
             let destination = self.destination.as_ref().unwrap();
             WriteDestination::in_directory(destination)?
         };
 
-        let mut writer = ArtifactWriter::new(
-            &host_phase_2_id,
-            &host_phase_2_artifact.artifact,
-            &control_plane_id,
-            &control_plane_artifact.artifact,
-            destination,
-        );
-
-        writer.write(&log, &event_sender).await;
-
-        // TODO: verify artifact was correctly written out to disk.
-
-        // Drop the event sender: this signals completion.
-        _ = event_sender
-            .send(ReportEvent::Completion(CompletionEventKind::Completed))
-            .await;
-        std::mem::drop(event_sender);
+        // XXX: look at breaking this up into separate steps. This is tricky
+        // because of the combined logic around handling retries.
```

Comment:
Is this comment still relevant given `ArtifactWriter` itself uses a nested engine with separate steps?

Reply:
No, not any longer. Thanks!